docs: Added cookbook

vikram-dagger opened the pull request:
This commit adds a new cookbook page which is intended to aggregate code snippets from multiple sources, including existing documentation, GitHub repositories, Discord and others. Each recipe includes a brief explanation, code sample and optional link to source. Recipes are categorized within the page, although the categorization is currently fluid and may change.

gerhard reviewed (commented):
This content is great! I am keen to make it public ASAP.

A few improvement suggestions:

## 1. Cookbook**s**

This is a collection of cookbooks. Plural reads best IMO.

## 2. Better navigation between cookbooks

There are **a lot** of super useful cookbooks, but it is hard to understand this from the secondary navigation, the one on the right-hand side of the page.

How difficult would it be to use the same layout as Guides?

<img width="2878" alt="Screenshot 2023-04-14 at 13 47 06" src="https://user-images.githubusercontent.com/3342/232047647-c957c87f-ac64-4237-be72-26ef94677723.png">

The added benefit of this approach is that each cookbook would have its own canonical URL. When we use headings, it is more difficult to see which cookbook the link is for since multiple cookbooks might be visible at the same time on taller screens, e.g.:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/3342/232048479-1e907e8e-af4d-4ff1-beba-833a369c223c.png">

## 3. Highlight relevant lines

In the _Set host working directory_ cookbook, would it be possible to highlight the line which contains `dagger.WithWorkdir(".")`? It would make it easier to distinguish between the important line(s) vs the boilerplate code.

## 4. Go ~~1.19~~ 1.20

Our Go SDK requires Go 1.20 as of [`sdk/go/v0.5.0`](https://github.com/dagger/dagger/releases/tag/sdk%2Fgo%2Fv0.5.0). It is most likely that all examples with go 1.19 will not work.

## 5. Wording

Feel free to implement the suggestions which you like, ignore the rest:
- _Perform multi-stage build_ -> **Multi-stage build**
- _Perform matrix build_ -> **Matrix build**
- _Build multi-architecture image (emulation)_ -> **Multi-arch build**
- _Build multi-architecture image (cross-compilation)_ -> **Multi-arch build with cross-compilation**
- _Bind transient service_ -> **Use db for integration test**
- _Publish container_ -> **Publish container to Docker Hub**
- _Export container to host_ -> **Export container image to host**
- _Export directory from container to host_ -> **Export container directory to host**
- _Inject secret as environment variable_ -> **Expose secret via environment variable**
- _Inject secret as file_ -> **Expose secret via file**
- _Cache service state_ -> **Persist service state between runs**

vikram-dagger commented:
Thanks for the review @gerhard !

> A few improvement suggestions:
> ## 1. Cookbook**s**
> This is a collection of cookbooks. Plural reads best IMO.

AFAIK the general convention for code is singular ie a "cookbook" with different "chapters" eg. https://www.puppetcookbook.com/ and https://www.oreilly.com/library/view/python-cookbook-3rd/9781449357337/

Would it be better if we named the page differently eg.  "Code Library", "Code Samples", "Examples"?

Would like feedback from others on this as well!

> ## 2. Better navigation between cookbooks
> There are **a lot** of super useful cookbooks, but it is hard to understand this from the secondary navigation, the one on the right-hand side of the page.
> 
> How difficult would it be to use the same layout as Guides?

The Guides page is created using a custom React UX component, which generates the index page dynamically by looking for markdown files in a specific directory and/or with specific metadata (@crjm to confirm). For the cookbook, the code samples are in different locations in the docs tree and don't have any specific identifying metadata. So we can't use the guides UX directly, we would need to create a new UX component or adapt the existing one, and probably have additional metadata to configure it.

> ## 3. Highlight relevant lines
> In the _Set host working directory_ cookbook, would it be possible to highlight the line which contains `dagger.WithWorkdir(".")`? It would make it easier to distinguish between the important line(s) vs the boilerplate code.

Yes, this is one that I already attempted. In fact, some of the code samples already had highlights and I removed them. The issue here is that the code samples "belong" to other guides and are already in use in those guides - the cookbook page is merely a different view that reuses (not duplicates) them. So, adding highlights for a specific code sample may make sense in cookbook context, but is confusing when the same highlights are seen in the context of the original guide (and vice-versa as well). However, I 100% agree this would be a useful feature to have, maybe @crjm can suggest a way to make it work while still maintaining the reusability.

> ## 4. Go ~1.19~ 1.20
> Our Go SDK requires Go 1.20 as of [`sdk/go/v0.5.0`](https://github.com/dagger/dagger/releases/tag/sdk%2Fgo%2Fv0.5.0). It is most likely that all examples with go 1.19 will not work.

Can you please point me to specific cases where you found this issue?

> ## 5. Wording
> Feel free to implement the suggestions which you like, ignore the rest:

I implemented the majority of these, thank you!

gerhard commented:
1. Yes, `Cookbook`, no plural 👍
2. OK, we may need some help from @crjm or @slumbering on this one.
3. Sounds like another follow-up improvement to me, no worries.
4. Everything under the **Integrations** section which references Go 1.19 should be updated to 1.20 - e.g. `go-version: 1.19` -> `go-version: '1.20'`
5. 🙌

gerhard reviewed (approved):
This is too good to not ship this Friday. Go for merge & remember to publish it live. We can follow-up with the rest of the improvements next week.

vikram-dagger commented:
> 1. Yes, `Cookbook`, no plural +1

OK

>     4. Everything under the **Integrations** section which references Go 1.19 should be updated to 1.20 - e.g. `go-version: 1.19` -> `go-version: '1.20'`

Done

vikram-dagger commented:
> > This is too good to not ship this Friday. Go for merge & remember to publish it live. We can follow-up with the rest of the improvements next week.
> 
> Thanks! On publication and timing, I'd like to have @mircubed and @shykes opinions eg. if we want to tweet it, maybe it's optimal to publish it at the beginning of the week instead?

Merging and going live per discussion with Gerhard: https://discord.com/channels/@me/1027171337866846299/1097560367879704698
